File: pytests/helpers/sign.py
"""
sign.py — Binance 风格 HMAC-SHA256 签名助手（Python 版）
============================================================================
与 src/modules/pre-signer.ts、tests/helpers/sign.ts 的签名规则完全一致：
    signature = HMAC_SHA256(secret, 按参数名字母序拼接的 query string)
三套栈（k6 / Playwright / Pytest）对靶机的验签口径统一。
"""
import hashlib
import hmac
import os
import time

API_KEY = os.environ.get("API_KEY", "test-api-key")
API_SECRET = os.environ.get("API_SECRET", "test-api-secret")

SAMPLE_ORDER = {
    "symbol": "BTCUSDT",
    "side": "BUY",
    "type": "LIMIT",
    "timeInForce": "GTC",
    "quantity": 0.01,
    "price": 65000,
}


def sign_order_query(order, recv_window=5000, timestamp=None, secret=None):
    """生成排序后的 query 及其签名，返回 (query, signature)。"""
    params = dict(order)
    params["recvWindow"] = recv_window
    params["timestamp"] = timestamp if timestamp is not None else int(time.time() * 1000)

    query = "&".join(f"{k}={params[k]}" for k in sorted(params.keys()))
    signature = hmac.new(
        (secret or API_SECRET).encode(), query.encode(), hashlib.sha256
    ).hexdigest()
    return query, signature

File: tests/helpers/sign.ts
/**
 * sign.ts — Binance 风格 HMAC-SHA256 签名助手（Node crypto，供 E2E 下单用例复用）
 * ============================================================================
 * 与 src/modules/pre-signer.ts 的签名规则完全一致：
 *   signature = HMAC_SHA256(secret, 按参数名字母序拼接的 query string)
 * 使 Playwright 功能测试与 k6 压测使用相同的签名约定，靶机验签统一。
 */
import crypto from 'crypto';

export const apiKey = process.env.API_KEY || 'test-api-key';
const apiSecret = process.env.API_SECRET || 'test-api-secret';

export interface OrderParams {
  symbol: string;
  side: string;
  type: string;
  timeInForce: string;
  quantity: number;
  price: number;
}

export interface SignOpts {
  recvWindow?: number;
  timestamp?: number;
  secret?: string;
}

/** 生成排序后的 query 及其签名 */
export function signOrderQuery(p: OrderParams, opts: SignOpts = {}): { query: string; signature: string } {
  const params: Record<string, string | number> = {
    symbol: p.symbol,
    side: p.side,
    type: p.type,
    timeInForce: p.timeInForce,
    quantity: p.quantity,
    price: p.price,
    recvWindow: opts.recvWindow ?? 5000,
    timestamp: opts.timestamp ?? Date.now(),
  };
  const query = Object.keys(params)
    .sort()
    .map((k) => `${k}=${params[k]}`)
    .join('&');
  const signature = crypto.createHmac('sha256', opts.secret ?? apiSecret).update(query).digest('hex');
  return { query, signature };
}

export const SAMPLE_ORDER: OrderParams = {
  symbol: 'BTCUSDT',
  side: 'BUY',
  type: 'LIMIT',
  timeInForce: 'GTC',
  quantity: 0.01,
  price: 65000,
};
